internal/provider/base_eda_datasource.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	tfpath "github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ datasource.DataSource                     = &BaseEdaDataSource{}
+	_ datasource.DataSourceWithConfigure        = &BaseEdaDataSource{}
+	_ datasource.DataSourceWithConfigValidators = &BaseEdaDataSource{}
+	_ datasource.DataSourceWithValidateConfig   = &BaseEdaDataSource{}
+)
+
+// NewBaseEdaDataSource creates a new `BaseEdaDataSource`.
+func NewBaseEdaDataSource(client ProviderHTTPClient, stringDescriptions StringDescriptions) *BaseEdaDataSource {
+	return &BaseEdaDataSource{
+		client:             client,
+		StringDescriptions: stringDescriptions,
+	}
+}
+
+// Metadata returns the data source type name composing it from the provider type name and the
+// entity slug string passed in the constructor.
+func (d *BaseEdaDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_%s", req.ProviderTypeName, d.MetadataEntitySlug)
+	fmt.Println(resp.TypeName)
+}
+
+// Schema defines the schema fields for the data source.
+func (d *BaseEdaDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"name": schema.StringAttribute{
+				Required: true,
+			},
+			"url": schema.StringAttribute{
+				Computed: true,
+			},
+			"id": schema.Int64Attribute{
+				Computed: true,
+			},
+		},
+	}
+}
+
+// ConfigValidators validates configuration in a declarative way.
+func (d *BaseEdaDataSource) ConfigValidators(_ context.Context) []datasource.ConfigValidator {
+	// You have at least an name
+	return []datasource.ConfigValidator{
+		datasourcevalidator.Any(
+			datasourcevalidator.AtLeastOneOf(
+				tfpath.MatchRoot("name")),
+		),
+	}
+}
+
+// ValidateConfig validates configuration in an imperative way.
+func (d *BaseEdaDataSource) ValidateConfig(ctx context.Context, req datasource.ValidateConfigRequest, resp *datasource.ValidateConfigResponse) {
+	// Check that the response and diagnostics pointer is defined
+	if resp == nil {
+		tflog.Error(ctx, "Response not defined, we cannot continue with the execution")
+		return
+	}
+
+	// Check that the current context is active
+	if !IsContextActive("ValidateConfig", ctx, &resp.Diagnostics) {
+		return
+	}
+
+	var data BaseEdaSourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if IsValueProvidedOrPromised(data.Name) {
+		return
+	}
+
+	if !IsValueProvidedOrPromised(data.Name) {
+		resp.Diagnostics.AddAttributeWarning(
+			tfpath.Root("id"),
+			"Missing Attribute Configuration",
+			"Expected [id]",
+		)
+	}
```

Comment:
I had a couple questions but first let me share my understanding of this code.

It looks to me like we first read a `BaseEdaSourceModel` from `req.Config` (no confusion there). Then, we check if that data has a **Name** (`IsValueProvidedOrPromised(data.Name)`).

If it does, 👍 . If not, we add a warning about a missing **id** (not name) 🤔.

So my first question is - why do we need to do that? Does the schema not handle this? Second question - if we do, should this be `name` and not `id`?

As an exercise, I'd be interested to see the output of `terraform validate` on the following snippet with all of these validators removed:

```hcl
data.aap_eda_eventstream {
  # empty on purpose
}
```

Reply:
Good catch! This was a copy/paste error and I didn't update the warning text. I'll make sure I do that.

Reply:
I've removed the `ConfigValidators` and `ValidateConfig` functions, which also removed this code.